File: packages/contracts/go/invoice.go
// GST invoice and invoice numbering — added at 0.4.0 (ADR-016, Milestone 3).
// Mirrors src/types/invoice.ts.
//
// Invoice is EDGE-AUTHORITATIVE (§50.1): the outlet issues bills with the
// uplink down, and the cloud only ever REPLAYS them. No handler in this
// backend mints an invoice number or transitions an invoice — the rule ADR-014
// set for kot.status, applied to money.
//
// InvoiceSeries is CONFIG (cloud→edge): the series *definition*. The counter
// that produces the next number is edge-local (sqlite `invoice_sequence`),
// deliberately has no Go struct and no Postgres table, and never syncs.
// Mirroring it would make the cloud a second writer of invoice numbers, which
// §33's "never generate duplicate invoice numbers" forbids.
package contracts

import "time"

type InvoiceStatus string

const (
	InvoiceStatusIssued    InvoiceStatus = "ISSUED"
	InvoiceStatusCancelled InvoiceStatus = "CANCELLED"
)

type SequenceResetPolicy string

const (
	SequenceResetNever SequenceResetPolicy = "NEVER"
	SequenceResetFY    SequenceResetPolicy = "FY"
	SequenceResetMonth SequenceResetPolicy = "MONTH"
	SequenceResetDay   SequenceResetPolicy = "DAY"
)

type InvoiceSeries struct {
	ID       string `json:"id"`
	OutletID string `json:"outlet_id"`
	Code     string `json:"code"`
	// Tokens: {FY} {YYYY} {MM} {DD} {OUTLET}. 'FY{FY}/{OUTLET}/' with
	// PaddingWidth 6 renders FY26/PNQ/001423.
	PrefixTemplate string              `json:"prefix_template"`
	ResetPolicy    SequenceResetPolicy `json:"reset_policy"`
	PaddingWidth   int                 `json:"padding_width"`
	IsActive       bool                `json:"is_active"`
	ConfigVersion  int                 `json:"config_version"`
	SchemaVersion  int                 `json:"schema_version"`
}

// TaxLiabilityParty records who bears the GST liability (§32). Captured at
// issue time because direct and ECO supplies must never be combined in
// compliance reporting, and that is only possible if the classification was
// recorded when the bill was raised. Milestone 3 EXCLUDES the reporting
// outputs, not these fields.
type TaxLiabilityParty string

const (
	TaxLiabilityRestaurant TaxLiabilityParty = "RESTAURANT"
	TaxLiabilityECO        TaxLiabilityParty = "ECO"
)

type InvoiceLine struct {
	ID        string `json:"id"`
	InvoiceID string `json:"invoice_id"`
	// The order line this bills. This is what makes the split-bill conservation
	// property checkable: across a split group every order line must appear
	// exactly once in total quantity — no loss, no duplication, no double-tax.
	OrderItemID string `json:"order_item_id"`
	LineNo      int    `json:"line_no"`
	// Snapshot at issue time — never re-read from the live menu.
	Description       string  `json:"description"`
	HSNSAC            *string `json:"hsn_sac"`
	Quantity          int     `json:"quantity"`
	UnitPricePaise    int     `json:"unit_price_paise"`
	GrossPaise        int     `json:"gross_paise"`
	DiscountPaise     int     `json:"discount_paise"`
	TaxableValuePaise int     `json:"taxable_value_paise"`
	TaxProfileID      string  `json:"tax_profile_id"`
	CGSTRateBps       int     `json:"cgst_rate_bps"`
	CGSTPaise         int     `json:"cgst_paise"`
	SGSTRateBps       int     `json:"sgst_rate_bps"`
	SGSTPaise         int     `json:"sgst_paise"`
	IGSTRateBps       int     `json:"igst_rate_bps"`
	IGSTPaise         int     `json:"igst_paise"`
	CessRateBps       int     `json:"cess_rate_bps"`
	CessPaise         int     `json:"cess_paise"`
	TotalPaise        int     `json:"total_paise"`
	SchemaVersion     int     `json:"schema_version"`
}

// Invoice is a GST invoice (§33). Split bills are N invoices over one order
// sharing a SplitGroupID — each part is a real, independently numbered,
// independently payable invoice, because that is what the customer physically
// receives. There is deliberately no BillSplit entity.
type Invoice struct {
	ID       string `json:"id"`
	OutletID string `json:"outlet_id"`
	OrderID  string `json:"order_id"`

	SplitGroupID *string `json:"split_group_id"`
	SplitIndex   int     `json:"split_index"`
	SplitCount   int     `json:"split_count"`

	SeriesID      string    `json:"series_id"`
	InvoiceNumber string    `json:"invoice_number"`
	InvoiceDate   time.Time `json:"invoice_date"`
	// Outlet-local YYYY-MM-DD; the business day may cross midnight (CLAUDE.md).
	BusinessDate string `json:"business_date"`

	Status          InvoiceStatus `json:"status"`
	CancelledReason *string       `json:"cancelled_reason"`
	CancelledAt     *time.Time    `json:"cancelled_at"`

	CustomerName           *string `json:"customer_name"`
	CustomerPhone          *string `json:"customer_phone"`
	CustomerGSTIN          *string `json:"customer_gstin"`
	PlaceOfSupplyStateCode string  `json:"place_of_supply_state_code"`

	Lines []InvoiceLine `json:"lines"`

	// Money — integer paise throughout (CLAUDE.md §Money).
	SubtotalPaise     int `json:"subtotal_paise"`
	DiscountPaise     int `json:"discount_paise"`
	TaxableValuePaise int `json:"taxable_value_paise"`
	CGSTPaise         int `json:"cgst_paise"`
	SGSTPaise         int `json:"sgst_paise"`
	IGSTPaise         int `json:"igst_paise"`
	CessPaise         int `json:"cess_paise"`
	RoundOffPaise     int `json:"round_off_paise"`
	GrandTotalPaise   int `json:"grand_total_paise"`

	// Reproducibility (§31): the resolved rules AND the seller identity as they
	// stood at issue time, so a reprint after a rate or GSTIN change produces
	// the original document rather than a recomputed one.
	ComplianceVersionID string                 `json:"compliance_version_id"`
	TaxSnapshot         map[string]interface{} `json:"tax_snapshot"`
	FiscalProfile       map[string]interface{} `json:"fiscal_profile"`

	// ECO (§32) — modelled now, reported later.
	Channel              Channel           `json:"channel"`
	TaxLiabilityParty    TaxLiabilityParty `json:"tax_liability_party"`
	ECOOperatorName      *string           `json:"eco_operator_name"`
	ECOOperatorGSTIN     *string           `json:"eco_operator_gstin"`
	SupplyClassification *string           `json:"supply_classification"`

	CreatedByUserID string    `json:"created_by_user_id"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
	Version         int       `json:"version"`
	SchemaVersion   int       `json:"schema_version"`
}

// Channel is the sales channel an invoice was raised through. Kept as a named
// string rather than an enum because aggregator channels arrive in Milestone 6
// and the set is not closed at 0.4.0.
type Channel string

// SumsCorrectly reports whether the invoice satisfies the ADR-016 rounding
// policy: components sum to the grand total through round-off, round-off never
// exceeds half a rupee, and the total settles in whole rupees.
//
// The same policy is a CHECK in sqlite/0006 and postgres/0007 and a refine in
// the Zod schema. This method exists so an ingest handler can reject a
// malformed replay with a 422 that names the rule, rather than surfacing a raw
// constraint violation from the driver.
func (i Invoice) SumsCorrectly() bool {
	sum := i.TaxableValuePaise + i.CGSTPaise + i.SGSTPaise + i.IGSTPaise + i.CessPaise + i.RoundOffPaise
	if sum != i.GrandTotalPaise {
		return false
	}
	if i.RoundOffPaise > 50 || i.RoundOffPaise < -50 {
		return false
	}
	return i.GrandTotalPaise%100 == 0
}

File: packages/contracts/src/types/invoice.ts
// GST invoice and invoice numbering — added at 0.4.0 (ADR-016, Milestone 3).
// Mirrors sqlite/0006_m3_billing.sql and postgres/0007_m3_billing.sql.
//
// Invoice is EDGE-AUTHORITATIVE (§50.1): the outlet issues bills with the
// uplink down, and the cloud only ever replays them. No cloud handler mints an
// invoice number or transitions an invoice — the rule ADR-014 set for
// kot.status, applied to money.
//
// InvoiceSeries is CONFIG (cloud→edge): the series *definition*. The counter
// that produces the next number is edge-local (sqlite `invoice_sequence`),
// deliberately has no representation here, and never syncs. Splitting them is
// what keeps numbering concurrency-safe and offline-capable at once (§33).

import { z } from "zod";
import { RateBpsSchema } from "./tax";

export const InvoiceStatusSchema = z.enum(["ISSUED", "CANCELLED"]);
export type InvoiceStatus = z.infer<typeof InvoiceStatusSchema>;

export const SequenceResetPolicySchema = z.enum(["NEVER", "FY", "MONTH", "DAY"]);
export type SequenceResetPolicy = z.infer<typeof SequenceResetPolicySchema>;

export const InvoiceSeriesSchema = z.object({
  id: z.string().uuid(),
  outlet_id: z.string().uuid(),
  code: z.string(), // 'SALES', 'CREDIT_NOTE'
  // Tokens: {FY} {YYYY} {MM} {DD} {OUTLET}. 'FY{FY}/{OUTLET}/' with
  // padding_width 6 renders FY26/PNQ/001423 — the short human-facing format
  // CLAUDE.md §Money/time/identifiers requires.
  prefix_template: z.string(),
  reset_policy: SequenceResetPolicySchema,
  padding_width: z.number().int().min(1).max(12),
  is_active: z.boolean(),
  config_version: z.number().int(),
  schema_version: z.literal(1),
});
export type InvoiceSeries = z.infer<typeof InvoiceSeriesSchema>;

// Who bears the GST liability (§32). Modelled at issue time because direct and
// ECO supplies must never be combined in compliance reporting, and that is
// only possible if the classification was captured when the bill was raised.
// Milestone 3 EXCLUDES the reporting outputs, not these fields.
export const TaxLiabilityPartySchema = z.enum(["RESTAURANT", "ECO"]);
export type TaxLiabilityParty = z.infer<typeof TaxLiabilityPartySchema>;

export const InvoiceLineSchema = z.object({
  id: z.string().uuid(),
  invoice_id: z.string().uuid(),
  // The order line this bills. This is what makes the split-bill conservation
  // property checkable: across a split group, every order line must appear
  // exactly once in total quantity — no loss, no duplication, no double-tax.
  order_item_id: z.string().uuid(),
  line_no: z.number().int().positive(),
  description: z.string(), // snapshot at issue time — never re-read from live menu
  hsn_sac: z.string().nullable(),
  quantity: z.number().int().positive(),
  unit_price_paise: z.number().int(),
  gross_paise: z.number().int(),
  discount_paise: z.number().int().default(0),
  taxable_value_paise: z.number().int(),
  tax_profile_id: z.string().uuid(),
  cgst_rate_bps: RateBpsSchema.default(0),
  cgst_paise: z.number().int().default(0),
  sgst_rate_bps: RateBpsSchema.default(0),
  sgst_paise: z.number().int().default(0),
  igst_rate_bps: RateBpsSchema.default(0),
  igst_paise: z.number().int().default(0),
  cess_rate_bps: RateBpsSchema.default(0),
  cess_paise: z.number().int().default(0),
  total_paise: z.number().int(),
  schema_version: z.literal(1),
});
export type InvoiceLine = z.infer<typeof InvoiceLineSchema>;

export const InvoiceSchema = z
  .object({
    id: z.string().uuid(),
    outlet_id: z.string().uuid(),
    order_id: z.string().uuid(),

    // Split bills are N invoices over one order sharing a split_group_id. Each
    // part is independently numbered and independently payable, because that
    // is what the customer physically receives. There is no bill_split entity.
    split_group_id: z.string().uuid().nullable(),
    split_index: z.number().int().positive().default(1),
    split_count: z.number().int().positive().default(1),

    series_id: z.string().uuid(),
    invoice_number: z.string(),
    invoice_date: z.string().datetime(), // UTC storage (CLAUDE.md §Time)
    business_date: z.string(), // outlet-local YYYY-MM-DD; may cross midnight

    status: InvoiceStatusSchema,
    cancelled_reason: z.string().nullable(),
    cancelled_at: z.string().datetime().nullable(),

    customer_name: z.string().nullable(),
    customer_phone: z.string().nullable(),
    customer_gstin: z.string().nullable(),
    place_of_supply_state_code: z.string(),

    lines: z.array(InvoiceLineSchema),

    // Money — integer paise throughout (CLAUDE.md §Money).
    subtotal_paise: z.number().int(),
    discount_paise: z.number().int().default(0),
    taxable_value_paise: z.number().int(),
    cgst_paise: z.number().int().default(0),
    sgst_paise: z.number().int().default(0),
    igst_paise: z.number().int().default(0),
    cess_paise: z.number().int().default(0),
    round_off_paise: z.number().int().default(0),
    grand_total_paise: z.number().int(),

    // Reproducibility (§31). The resolved rules AND the seller identity as
    // they stood at issue time, so a reprint after a rate or GSTIN change
    // produces the original document rather than a recomputed one.
    compliance_version_id: z.string().uuid(),
    tax_snapshot: z.record(z.unknown()),
    fiscal_profile: z.record(z.unknown()),

    // ECO (§32) — modelled now, reported later.
    channel: z.string(),
    tax_liability_party: TaxLiabilityPartySchema,
    eco_operator_name: z.string().nullable(),
    eco_operator_gstin: z.string().nullable(),
    supply_classification: z.string().nullable(),

    created_by_user_id: z.string().uuid(),
    created_at: z.string().datetime(),
    updated_at: z.string().datetime(),
    version: z.number().int(),
    schema_version: z.literal(1),
  })
  // The ADR-016 rounding policy, stated a third time. It is already a CHECK in
  // sqlite/0006 and postgres/0007; repeating it here means a malformed bill
  // cannot even be constructed in TypeScript, let alone stored or replayed.
  // Tax is summed per component across the invoice and rounded half-up once;
  // the grand total is then rounded to the nearest rupee, with the delta in
  // round_off_paise.
  .refine(
    (inv) =>
      inv.grand_total_paise ===
      inv.taxable_value_paise +
        inv.cgst_paise +
        inv.sgst_paise +
        inv.igst_paise +
        inv.cess_paise +
        inv.round_off_paise,
    { message: "grand_total_paise must equal taxable value + tax components + round_off (ADR-016)" },
  )
  // Rounding to the nearest rupee can never move a total by more than half a
  // rupee. A larger value means round-off is absorbing an arithmetic error.
  .refine((inv) => Math.abs(inv.round_off_paise) <= 50, {
    message: "round_off_paise cannot exceed half a rupee (ADR-016)",
  })
  .refine((inv) => inv.grand_total_paise % 100 === 0, {
    message: "grand_total_paise must settle in whole rupees (ADR-016)",
  })
  .refine((inv) => inv.split_index <= inv.split_count, {
    message: "split_index cannot exceed split_count",
  });
export type Invoice = z.infer<typeof InvoiceSchema>;
